src/ripple/app/reporting/ETLHelpers.h

```diff
@@ -0,0 +1,194 @@
+    /// the datastructure has been stopped
+    std::optional<uint32_t>
+    getMostRecent()
+    {
+        std::unique_lock lck(mtx_);
+        cv_.wait(lck, [this]() { return max_ || stopping_; });
+        return max_;
+    }
+
+    /// Waits for the sequence to be validated by the network
+    /// @param sequence to wait for
+    /// @return true if sequence was validated, false otherwise
+    /// a return value of false means the datastructure has been stopped
+    bool
+    waitUntilValidatedByNetwork(uint32_t sequence)
+    {
+        std::unique_lock lck(mtx_);
+        cv_.wait(lck, [sequence, this]() {
+            return (max_ && sequence <= *max_) || stopping_;
+        });
+        return !stopping_;
+    }
+
+    /// Puts the datastructure in the stopped state
+    /// Future calls to this datastructure will not block
+    /// This operation cannot be reversed
+    void
+    stop()
+    {
+        std::lock_guard lck(mtx_);
+        stopping_ = true;
+        cv_.notify_all();
+    }
+};
+
+/// Generic thread-safe queue with an optional maximum size
+template <class T>
+class ThreadSafeQueue
+{
+    std::queue<T> queue_;
+
+    std::mutex m_;
+    std::condition_variable cv_;
+    std::optional<uint32_t> maxSize_;
+
+public:
+    /// @param maxSize maximum size of the queue. Calls that would cause the
+    /// queue to exceed this size will block until free space is available
+    explicit ThreadSafeQueue(uint32_t maxSize) : maxSize_(maxSize)
+    {
+    }
+
+    /// Create a queue with no maximum size
+    ThreadSafeQueue() = default;
+
+    /// @param elt element to push onto queue
+    /// if maxSize is set, this method will block until free space is available
+    void
+    push(T const& elt)
+    {
+        std::unique_lock lck(m_);
+        // if queue has a max size, wait until not full
+        if (maxSize_)
+            cv_.wait(lck, [this]() { return queue_.size() <= *maxSize_; });
+        queue_.push(elt);
+        cv_.notify_all();
+    }
+
+    /// @param elt element to push onto queue. elt is moved from
+    /// if maxSize is set, this method will block until free space is available
+    void
+    push(T&& elt)
+    {
+        std::unique_lock lck(m_);
+        // if queue has a max size, wait until not full
+        if (maxSize_)
+            cv_.wait(lck, [this]() { return queue_.size() <= *maxSize_; });
+        queue_.push(std::move(elt));
+        cv_.notify_all();
+    }
+
+    /// @return element popped from queue. Will block until queue is non-empty
+    T
+    pop()
+    {
+        std::unique_lock lck(m_);
+        cv_.wait(lck, [this]() { return !queue_.empty(); });
+        T ret = std::move(queue_.front());
+        queue_.pop();
+        // if queue has a max size, unblock any possible pushers
+        if (maxSize_)
+            cv_.notify_all();
+        return ret;
+    }
+};
+
+/// Convenience function for printing out basic ledger info
+inline std::string
+toString(LedgerInfo const& info)
```

Comment:
moved this to `ReportingETL.cpp` (though didn't rename)

Reply:
Since this isn't "the" `toString` function for `LedgerInfo`, I'd probably put it into it's own namespace (probably `detail`). I don't think I'd put it into an anonymous namespace, as it gets too confusing if someone does add a `toString` function for `LedgerInfo`.